text/0000-function_output_type_parameters.md

```diff
@@ -0,0 +1,348 @@
+let d = a * b;     // OK
+let e = a * b;     // ERR, a and b moved, caused by foo::T not known to be Copy
+
+let x: foo::T = foo() // OK
+let y: foo::T = 0     // ERR, foo::T not known to be an integer
+
+```
+
+This feature would interact with other function-related items in the same way
+as generics already do:
+
+- Works for standalone functions, associated functions or methods.
+- Not definable for closure literals, since they use concrete, inferred types.
+- The `Output` associated type of closure traits
+  for such a concrete function item would be simply defined in terms of the
+  abstract output types of the function item itself:
+
+```rust
+fn foo()<T: Clone> -> T where T: Mul { 42 }
+// from compiler generated:
+impl Fn<()> for foo {
+    type Output = foo::T;
+    // ...
+}
+
+let x: foo::T = foo();
+let y: foo::Output = x;        // OK
+let z: &Fn() -> foo::T = &foo; // OK (but not very useful)
+
+```
+
+## Conditional bounds
+
+Extend `where` bounds with the concept of "conditional bounds":
+
+```rust
+T: Foo if U: Bar
+```
+
+Semantically, they express that the bound on the lhs of `if` is only
+fulfilled if the bound on the rhs is.
+
+They can appear in any location where `where` clauses are legal,
+including in trait definitions and abstract return types:
+
+```rust
+trait Foo where Self: Add if Self: Clone { }
+
+fn foo<T>(t: T)<U> -> U
+    where U: Clone if T: Clone,
+          U: Foo,
+{ t }
+
+impl Foo for i32 { }    // OK
+let x = foo::<i32>(42); // OK
+let y = x.clone();      // OK
+let z = x + y;          // OK
+let e = x * y;          // ERR, foo<i32>::U not known to be Mul
+
+struct X;
+impl Foo for X { }   // OK
+let a = foo::<X>(X); // OK
+let b = a.clone();   // ERR, foo<X>::U is not Clone because X is not Clone
+let c = a + b;       // ERR, foo<X>::U is not Add because foo<X>::U is not Clone
+
+#[derive(Clone)]
+struct Y;
+impl Foo for Y { }   // ERR, Y needs to implement Add if it implements Clone.
+
+```
+
+The bound needs to apply if the condition is met. Eg, for the example above:
+
+- If a type implements `Clone` but not `Add`,
+  implementing `Foo` for it is an type error.
+- If a type does not implement `Clone`, then it does not matter
+  if the type implements `Add` or not, but a `Foo` bound will not know
+  about `Self: Add`.
+
+## Sugar for defining generic type parameters
+
+A problem with the existing generic syntax is
+that its often more verbose than the non-generic version of the API. Compare:
+
+```rust
+fn take_iter<I: Iterator>(i: &mut I) { ... }
+fn take_iter(i: &mut Iterator) { ... }
+```
+
+Adding additional output type parameters to functions will only increase this
+issue.
+
+This proposes two incremental syntax sugars for shrinking down the size of
+generic declarations:
+
+1. If a generic type parameter only appears in one location in a function
+   signature, it may be declared directly inline with the syntax
+   `type IDENT[: BOUNDS]`.
+   Example:
+   ```rust
```

Comment:
This code block (as well as the one in example 2) is rendering strangely, maybe add a new line and remove the space before the ticks?